server/src/main/java/org/elasticsearch/script/ScriptTermStats.java

```diff
@@ -0,0 +1,233 @@
+        this.terms = terms.toArray(new Term[0]);
+        this.termContextsSupplier = CachedSupplier.wrap(this::loadTermContexts);
+        this.postingsSupplier = CachedSupplier.wrap(this::loadPostings);
+        this.docFreqSupplier = CachedSupplier.wrap(this::loadDocFreq);
+        this.totalTermFreqSupplier = CachedSupplier.wrap(this::loadTotalTermFreq);
+    }
+
+    /**
+     * Number of unique terms in the query.
+     *
+     * @return the number of unique terms
+     */
+    public int uniqueTermsCount() {
+        return terms.length;
+    }
+
+    /**
+     * Number of terms that are matched im the query.
+     *
+     * @return the number of matched terms
+     */
+    public int matchedTermsCount() {
+        final int docId = docIdSupplier.get();
+        int matchedTerms = 0;
+
+        try {
+            for (PostingsEnum postingsEnum : postingsSupplier.get()) {
+                if (postingsEnum != null && postingsEnum.advance(docId) == docId && postingsEnum.freq() > 0) {
+                    matchedTerms++;
+                }
+            }
+            return matchedTerms;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Collect docFreq (number of documents a term occurs in) for the terms of the query and returns statistics for them.
+     *
+     * @return statistics on docFreq for the terms of the query.
+     */
+    public StatsAccumulator docFreq() {
+        return docFreqSupplier.get();
+    }
+
+    private StatsAccumulator loadDocFreq() {
+        StatsAccumulator docFreqStats = new StatsAccumulator();
+        TermStates[] termContexts = termContextsSupplier.get();
+
+        try {
+            for (int i = 0; i < termContexts.length; i++) {
+                try {
+                    docFreqStats.accept(termStatistics(terms[i], termContexts[i]).docFreq());
+                } catch (IllegalArgumentException e) {
+                    docFreqStats.accept(termContexts[i].docFreq());
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return docFreqStats;
+    }
+
+    /**
+     * Collect totalTermFreq (total number of occurrence of a term in the index) for the terms of the query and returns statistics for them.
+     *
+     * @return statistics on totalTermFreq for the terms of the query.
+     */
+    public StatsAccumulator totalTermFreq() {
+        return this.totalTermFreqSupplier.get();
+    }
+
+    private StatsAccumulator loadTotalTermFreq() {
+        StatsAccumulator totalTermFreqStats = new StatsAccumulator();
+        TermStates[] termContexts = termContextsSupplier.get();
+
+        try {
+            for (int i = 0; i < termContexts.length; i++) {
+                try {
+                    totalTermFreqStats.accept(termStatistics(terms[i], termContexts[i]).totalTermFreq());
+                } catch (IllegalArgumentException e) {
```

Comment:
`termStatistics` is using the `Searcher::termStatistics` method:

```
private TermStatistics termStatistics(Term term, TermStates termStates) throws IOException {
    return searcher.termStatistics(term, termStates.docFreq(), termStates.totalTermFreq());
}
When a term is not present in the current shards, this is equivalent to:

searcher.termStatistics(term, 0, 0);
which uses aggregated Document Frequency Statistics (DFS) to retrieve term statistics:

if (aggregatedDfs == null) {
    // Either the DFS phase is executing, or the search type does not include a DFS phase.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

TermStatistics termStatistics = aggregatedDfs.termStatistics().get(term);
if (termStatistics == null) {
    // No stats available for this term - it could be from a "must_not" clause or similar, 
    // preventing term extraction from the query.
    return super.termStatistics(term, docFreq, totalTermFreq);
}

return termStatistics;
```

when a term is missing it will try to create a `TermStatistics` object with a docFreq  equal to 0, causing the IAE to be raised.

Because we want to leverage the stats from dfs for more reliable results in the distributed use case, we have to catch this edge case since there is no way to predict when the IAE will be thrown.

Reply:
This seems fragile. There could be other cases (for example, an actual bad argument, like reversing term freq and total term freq) for which an IAE is thrown. I think we should find a way to avoid calling termStatistics. For example, you could have a new method on ContextIndexSearcher, to find the totalTermFreq for a term (which takes a fallback). So something like:

```
if (searcher instanceof ContextIndexSearcher dfsSearcher) {
    totaltermFreqStats.accept(dfsSearcher.getTotalTermFreq(term, totalTermFreq));
} else {
    totaltermFreqStats.accept(totalTermFreq);
}
```

Reply:
I like your proposal and implemented something very similar in the last version I just pushed.